File: protocol/python/g8e_protocol/paths.py
# Copyright (c) 2026 Lateralus Labs, LLC.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
from pathlib import Path

def resolve_project_root() -> Path:
    """
    Resolves the project root directory.
    This is the canonical root detection heuristic - all languages must match this logic.
    Priority:
    1. G8E_PROJECT_ROOT environment variable
    2. Walk up from current directory looking for marker: services/ directory AND g8e file
    3. If in services/g8eo, walk up 2 levels to root
    4. If in services/g8ee, walk up 2 levels to root
    5. Fallback to current directory
    """
    env_root = os.environ.get('G8E_PROJECT_ROOT')
    if env_root:
        return Path(env_root).resolve()

    cwd = Path.cwd()

    # Try to find root by looking for the marker: services/ directory AND g8e file
    for parent in [cwd] + list(cwd.parents):
        if (parent / "services").exists() and (parent / "g8e").exists():
            return parent.resolve()

    # If in services/g8eo, walk up 2 levels to root
    if "services/g8eo" in str(cwd):
        for parent in [cwd] + list(cwd.parents):
            if parent.name == "g8eo" and parent.parent.name == "services":
                return parent.parent.parent.resolve()

    # If in services/g8ee, walk up 2 levels to root
    if "services/g8ee" in str(cwd):
        for parent in [cwd] + list(cwd.parents):
            if parent.name == "g8ee" and parent.parent.name == "services":
                return parent.parent.parent.resolve()

    # Fallback to current directory
    return cwd.resolve()

PROJECT_ROOT = resolve_project_root()
G8E_DIR = PROJECT_ROOT / ".g8e"

def resolve_path(env_var: str, default: Path) -> Path:
    """Resolve a path from an environment variable or a default.
    
    If the environment variable contains a relative path, it is resolved
    relative to the project root.
    """
    val = os.environ.get(env_var)
    if val:
        p = Path(val)
        if p.is_absolute():
            return p
        return (PROJECT_ROOT / p).resolve()
    return default.resolve()

PKI_DIR = resolve_path("G8E_PKI_DIR", G8E_DIR / "pki")
SECRETS_DIR = resolve_path("G8E_SECRETS_DIR", G8E_DIR / "secrets")
TRUST_BUNDLE_PATH = PKI_DIR / "trust" / "hub-bundle.pem"

def get_trust_bundle() -> str:
    """Return the path to the hub trust bundle.
    
    Resolution order:
    1. G8E_TRUST_BUNDLE (explicit override)
    2. PKI_DIR/trust/hub-bundle.pem
    
    Raises FileNotFoundError if no bundle is available.
    """
    explicit = os.environ.get("G8E_TRUST_BUNDLE", "").strip()
    if explicit:
        path = Path(explicit)
        if not path.exists():
            raise FileNotFoundError(f"G8E_TRUST_BUNDLE points to a missing file: {path}")
        return str(path)

    if not TRUST_BUNDLE_PATH.exists():
        raise FileNotFoundError(
            f"Hub trust bundle not found at {TRUST_BUNDLE_PATH}. "
            "Run `./g8e platform start` to provision PKI, or set G8E_TRUST_BUNDLE."
        )
    return str(TRUST_BUNDLE_PATH)

File: services/g8eo/internal/constants/paths.go
// Copyright (c) 2026 Lateralus Labs, LLC.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package constants

import (
	"os"
	"path/filepath"
)

// Paths defines canonical G8E filesystem paths.
var Paths = struct {
	Infra struct {
		DbPath               string
		PkiDir               string
		SecretsDir           string
		CaCertPath           string
		AppCertDir           string
		DocsDir              string
		ProtocolDir          string
		ProtocolConstantsDir string
		ProtocolModelsDir    string
		SshConfigPath        string
	}
}{
	Infra: struct {
		DbPath               string
		PkiDir               string
		SecretsDir           string
		CaCertPath           string
		AppCertDir           string
		DocsDir              string
		ProtocolDir          string
		ProtocolConstantsDir string
		ProtocolModelsDir    string
		SshConfigPath        string
	}{
		DbPath:               ".g8e/data/g8e.db",
		PkiDir:               ".g8e/pki",
		SecretsDir:           ".g8e/secrets",
		CaCertPath:           ".g8e/pki/trust/hub-bundle.pem",
		AppCertDir:           ".g8e/pki/issued/apps",
		DocsDir:              "/docs",
		ProtocolDir:          "/app/protocol",
		ProtocolConstantsDir: "/app/protocol/constants",
		ProtocolModelsDir:    "/app/protocol/models",
		SshConfigPath:        "/etc/g8e/ssh_config",
	},
}

func init() {
	resolvePaths()
}

// resolvePaths dynamically resolves filesystem paths from environment variables.
// Priority:
// 1. Explicit environment variable (e.g., G8E_PROTOCOL_DIR)
// 2. Relative to G8E_PROJECT_ROOT if set
// 3. Fallback to hardcoded default (container path)
func resolvePaths() {
	projectRoot := os.Getenv("G8E_PROJECT_ROOT")
	if projectRoot == "" {
		projectRoot = resolveProjectRoot()
	}

	// Resolve ProtocolDir
	if protocolDir := os.Getenv("G8E_PROTOCOL_DIR"); protocolDir != "" {
		if filepath.IsAbs(protocolDir) {
			Paths.Infra.ProtocolDir = protocolDir
		} else {
			Paths.Infra.ProtocolDir = filepath.Join(projectRoot, protocolDir)
		}
	} else {
		// Default to protocol/ relative to project root for host-native execution
		Paths.Infra.ProtocolDir = filepath.Join(projectRoot, "protocol")
	}

	// Update derived paths
	Paths.Infra.ProtocolConstantsDir = filepath.Join(Paths.Infra.ProtocolDir, "constants")
	Paths.Infra.ProtocolModelsDir = filepath.Join(Paths.Infra.ProtocolDir, "models")
}

// resolveProjectRoot returns the project root directory.
// This mirrors the logic in services/g8eo/internal/services/system/path.go
// but is duplicated here to avoid circular dependencies during init.
func resolveProjectRoot() string {
	if root := os.Getenv("G8E_PROJECT_ROOT"); root != "" {
		abs, err := filepath.Abs(root)
		if err == nil {
			return abs
		}
		return root
	}

	cwd, err := os.Getwd()
	if err != nil {
		return "/home/g8e"
	}

	// Try to find the root by looking for services/g8eo or .git
	current := cwd
	for {
		if _, err := os.Stat(filepath.Join(current, "services")); err == nil {
			if _, err := os.Stat(filepath.Join(current, "g8e")); err == nil {
				return current
			}
		}

		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
	}

	// Fallback to relative path from services/g8eo
	if contains(cwd, filepath.Join("services", "g8eo")) {
		current = cwd
		for {
			if filepath.Base(current) == "g8eo" && filepath.Base(filepath.Dir(current)) == "services" {
				return filepath.Dir(filepath.Dir(current))
			}
			parent := filepath.Dir(current)
			if parent == current {
				break
			}
			current = parent
		}
	}

	return cwd
}

// contains checks if a string contains a substring (Go < 1.21 compatibility)
func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(s) > len(substr) && (s[:len(substr)] == substr || s[len(s)-len(substr):] == substr || containsMiddle(s, substr)))
}

func containsMiddle(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}
